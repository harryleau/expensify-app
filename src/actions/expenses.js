import database from '../firebase/firebase';

//////// NO FIREBASE 
// component calls action generator
// action generator returns object
// component calls action generator
// redux store changes

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

//////// WITH FIREBASE
// component calls action generator
// action generator returns function
// component dispatches function - startAddExpense returns a function that dispatches action addExpense
// function runs (has the ability to dispatch other actions and do whatever it wants)

// START_ADD_EXPENSE
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
    const expense = { description, note, amount, createdAt };

    // for further promise chaining, we use return. for example, in test file, we can chain assertion after calling startAddExpense
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});