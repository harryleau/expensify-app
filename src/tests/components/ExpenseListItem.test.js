import React from 'react';
import { shallow } from 'enzyme';
import ExpenselistItem from './../../components/ExpenseListItem';
import expenses from './../fixtures/expenses';
import ExpenseListItem from './../../components/ExpenseListItem';

test('should render ExpenseListItem with fixture data', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});