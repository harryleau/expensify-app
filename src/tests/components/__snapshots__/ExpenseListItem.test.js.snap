// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with fixture data 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Book
    </h3>
  </Link>
  <p>
    2095
     - 
    -345600000
  </p>
</div>
`;
